plugins/amd-gpu/src/amd/device.rs

```diff
@@ -0,0 +1,129 @@
+                        let bus_id = amdsmi_get_gpu_device_bdf(gpu)
+                            .map_err(AmdError)
+                            .context("failed to get the bus ID of device")?;
+
+                        if features.has_any() {
+                            // Extract the device pointer because we will manage the lifetimes ourselves.
+                            Some(ManagedDevice {
+                                handle: gpu,
+                                features,
+                                bus_id: bus_id.to_string(),
+                            })
+                        } else {
+                            log::warn!("Skipping GPU device ({bus_id}) because it supports no useful feature.");
+                            None
+                        }
+                    }
+                    Err(e) => {
+                        if skip_failed_devices {
+                            log::warn!("Skipping GPU device because of error:\n{e:?}");
+                            None
+                        } else {
+                            // don't skip, fail immediately
+                            Err(AmdError(e))?
+                        }
+                    }
+                };
+                devices.push(device);
+            }
+        }
+        Ok(AmdGpuDevices { devices })
+    }
+
+    /// Gets and return status of GPU device detection on the operating system.
+    pub fn detection_stats(&self) -> DetectionStats {
+        let n_found = self.devices.len();
+        let n_failed = self.devices.iter().filter(|d| d.is_none()).count();
+        let n_ok = n_found - n_failed;
+        DetectionStats {
+            found_devices: n_found,
+            working_devices: n_ok,
+        }
+    }
+}
+
+#[cfg(test)]
+mod tests {
+    use super::*;
+    use amdsmi::{amdsmi_init, amdsmi_shut_down, amdsmi_get_socket_handles, AmdsmiInitFlagsT::AmdsmiInitAmdGpus};
+
+    // Test `detection_stats` function with features available
+    #[test]
+    fn test_detect_skips_device_with_features() {
+        let devices = vec![
+            Some(ManagedDevice {
+                handle: std::ptr::null_mut(),
+                features: OptionalFeatures::default(),
+                bus_id: "0000:00:01.0".to_string(),
+            }),
+        ];
+
+        let amd = AmdGpuDevices { devices };
+        let stats = amd.detection_stats();
+        assert_eq!(stats.found_devices, 1);
+        assert_eq!(stats.working_devices, 1);
+    }
+
+    // Test `detect` function with existing and detected devices
+    #[test]
+    fn test_detect_existing_devices() {
+        amdsmi_init(AmdsmiInitAmdGpus).map_err(AmdError).context("Failed to initialize AMD SMI");
+
+        match AmdGpuDevices::detect(true) {
+            Ok(devices) => {
+                let res = devices.detection_stats();
+                assert!(res.found_devices >= 0);
+            }
+            Err(e) => {
+                assert!(false);
+            }
+        }
```

Comment:
Things like that should be written:
```suggestion
        let devices = AmdGpuDevices::detect(true).expect("detection failure");
        assert!(devices.detection_stats().found_devices >= 0);
```

Reply:
Yes and also this test is usless now, it's already cover by the integration test.